### PR TITLE
HDDS-8822. [S3G] Improve list performance in LEGACY/OBS bucket

### PR DESCRIPTION
**Pull request by whbing**
## What changes were proposed in this pull request?

[S3G] Improve list performance in LEGACY/OBS bucket when listing delimited by '/'

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8822

## How was this patch tested?

- Manual test

### (1) LEGACY bucket list optimization
keys info:
```bash
hadoop fs -count ofs://om/s3v/buk-lg/*
           6        54712             109424 ofs://om/s3v/buk-lg/s3put
           1            1                238 ofs://om/s3v/buk-lg/test1
          61        30489              30489 ofs://om/s3v/buk-lg/test2
          21       100000             100000 ofs://om/s3v/buk-lg/test3

hadoop fs -count ofs://om/s3v/buk-lg/test3/*
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir0_47640
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir10_19047
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir11_99507
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir12_43926
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir13_98700
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir14_30007
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir15_32065
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir16_42162
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir17_71071
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir18_46050
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir19_86567
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir1_78108
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir2_53468
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir3_60271
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir4_41931
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir5_92207
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir6_93526
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir7_98836
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir8_71060
           1         5000               5000 ofs://om/s3v/buk-lg/test3/dir9_11147
```
time before and after optimization:
```bash
$ time aws s3api --endpoint http://<old-s3g-ip>:9878 list-objects-v2 --bucket buk-lg --prefix 't' --delimiter '/'
{
    "CommonPrefixes": [
        {
            "Prefix": "test1/"
        },
        {
            "Prefix": "test2/"
        },
        {
            "Prefix": "test3/"
        }
    ]
}

real	0m13.022s
user	0m0.952s
sys	0m0.161s

$ time aws s3api --endpoint http://<new-s3g-ip>:9878 list-objects-v2 --bucket buk-lg --prefix 't' --delimiter '/'
{
    "CommonPrefixes": [
        {
            "Prefix": "test1/"
        },
        {
            "Prefix": "test2/"
        },
        {
            "Prefix": "test3/"
        }
    ]
}

real	0m3.210s
user	0m0.969s
sys	0m0.169s
```
Detail data are shown in the following table:

 time with prefix     | t | test3/   | '' 
-------- | ----- | --- | -- 
before optimization  |  13.022s | 10.381s | 16.430s
after optimization  | 3.210s | 3.134s | 3.132s

### (2) OBS bucket list optimization
keys info:
```
aws s3api --endpoint http://<s3g-ip>:9878 list-objects-v2 --bucket buk-obs --prefix 'obs1/obs2/obs3/' --max-keys 100000 | grep Key | wc -l
9986
aws s3api --endpoint http://<s3g-ip>:9878 list-objects-v2 --bucket buk-obs --prefix 'obs1/obs2/obs31/' --max-keys 100000 | grep Key | wc -l
10002
aws s3api --endpoint http://<s3g-ip>:9878 list-objects-v2 --bucket buk-obs --prefix 'obs1/obs2/obs311/' --max-keys 100000 | grep Key | wc -l
57549
```
time before and after optimization:
```bash
$ time aws s3 --endpoint http://<old-s3g-ip>:9878 ls buk-obs/obs1/obs2/obs3
                           PRE obs3/
                           PRE obs31/
                           PRE obs311/

real	0m9.165s
user	0m0.919s
sys	0m0.114s
```
```bash
$ time aws s3 --endpoint http://<new-s3g-ip>:9878 ls buk-obs/obs1/obs2/obs3
                           PRE obs3/
                           PRE obs31/
                           PRE obs311/

real	0m3.124s
user	0m0.927s
sys	0m0.159s
```

Detail data are shown in the following table:

 time with prefix     | obs1/obs2/obs3 | obs1/obs2/   | '' 
-------- | ----- | --- | -- 
before optimization  |  9.165s | 9.112s | 9.045s
after optimization  | 3.124s | 3.067s | 3.014s 

Note：reduce `ozone.client.list.cache` to simulate multiple calls to iterators

### CODE REVIEW

**Comment by whbing** [comment]
Related to https://github.com/apache/ozone/pull/4868

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java [hunk 1]
```diff
@@ -1099,9 +1131,105 @@ public OzoneKey next() {
...
+     *      a/
+     *      a/b1/
+     *      a/b1/c1/
+     *      a/b1/c2/
+     *      a/b2/
+     *      a/b2/d1/
+     *      a/b2/d2/
+     *      a/b2/d2/d21.txt
+     *      a/b2/d3/
+     *      a/b2/d3/d22.txt
```
Comment (ivandika3) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java:1172]: Based on the above structure, is it supposed to be `a/b2/d2/d22.txt`?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java [hunk 2]
```diff
@@ -1099,9 +1131,105 @@ public OzoneKey next() {
...
+     *      a/b3/
+     *      a/b3/e1/
+     *      a/b3/e2/
+     *      a/b3/e3/
+     *      a/b3/e3/e11.txt
```
Comment (ivandika3) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java:1177]: Similarly, is this supposed to be `a/b3/e1/e11.txt`?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java (hunk 1 above)
Comment (whbing) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java]: > Based on the above structure, is it supposed to be `a/b2/d2/d22.txt`?

Thanks ! fixed.

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/client/OzoneBucketStub.java [hunk 3]
```diff
@@ -243,6 +245,45 @@ public Iterator<? extends OzoneKey> listKeys(String keyPrefix,
...
+          String newKeyName;
+          if (res.length < 2) {
+            newKeyName = key.getName();
+          } else if (res.length == 2) {
+            newKeyName = res[0] + OZONE_URI_DELIMITER + res[1];
+          } else {
+            newKeyName =
+                res[0] + OZONE_URI_DELIMITER + res[1] + OZONE_URI_DELIMITER;
+          }
+          return new OzoneKey(key.getVolumeName(),
```
Comment (captainzmc) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/client/OzoneBucketStub.java:270]: ditto.

**Review by captainzmc — COMMENTED** [review]
Thanks to @whbing for the patch, the change looks good. there some minor comments. In addition, CI should have other problems that need to be fixed.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java [hunk 4]
```diff
@@ -1099,9 +1132,121 @@ public OzoneKey next() {
...
+      // already no result matching the keyPrefix.
+      return statuses.stream()
+          .map(status -> {
+            OmKeyInfo keyInfo = status.getKeyInfo();
+            String keyName = keyInfo.getKeyName();
+            if (status.isDirectory()) {
+              // add trailing slash to represent directory
+              keyName = OzoneFSUtils.addTrailingSlashIfNeeded(keyName);
+            }
+            return new OzoneKey(keyInfo.getVolumeName(),
```
Comment (xichen01) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java:1241]: Constructing a `OzoneKey` from  `OmKeyInf` appears many times in the code, so maybe we can create a tool function to do those. (Such as line L1608 and L1508)

**Review by xichen01 — COMMENTED** [review]
The change looks good，just a comment you can refer to.

**Comment by whbing** [comment]
Sorry for the late follow-up. I added the test and method doc for `listKeys`, and also tested the s3 list for legacy and obs buckets in the test environment. @captainzmc   please help triger ci, thanks !

**Comment by captainzmc** [comment]
@whbing，Sorry for the late reply. CI has been triggered

**Comment by whbing** [comment]
> @whbing，Sorry for the late reply. CI has been triggered

@captainzmc Thanks! I will add more test cases if needed  and improve some other boundary cases in the coming days.

**Comment by whbing** [comment]
@captainzmc Thanks for review. I added a param `ozone-s3g.list-keys.shallow.enabled`  to compare the efficiency before and after using this feature. The time taken for both the `true` and `false` settings corresponds with the initial description provided on this page.

**Review by captainzmc — APPROVED** [review]
LGTM, Thanks @whbing for the update, the change looks good.Hi @xichen01 @ivandika3 Would you take another look?

**Comment by captainzmc** [comment]
Let's merge this, Thanks @whbing for the patch. And Thanks @ivandika3 @xichen01 for the review.

**Pull request merged**